README.md
Number to Text
===
[![Build Status](https://travis-ci.org/garora/number2text.svg?branch=master)](https://travis-ci.org/garora/number2text)   [![Coverage Status](https://coveralls.io/repos/garora/number2text/badge.svg)](https://coveralls.io/r/garora/number2text)
> Provides the facility to convert number to text. 

* Default type is Indian for non-currency 
* Default type is Indian Rupee for currency
* Indian style for number to text (Indian Rupee for currency)
* English style for number to text (US Dollar for currency)


## How to install?

```shell
  npm install number2text
```

## How to use?

```js
  var numberToText = require('number2text');
 ```

```js
     
  var textDefault = numberToText(100); //it converts to default type i.e. Indian

  console.log('Converts 1000000 to ', textDefault); //Converts 1000000 to Ten Lakh
```

```js
       
  var textIndian = numberToText(1000000,'Indian');

  console.log('Converts 1000000 to ', textIndian); //Converts 1000000 to Ten Lakh
```

```js
       
  var textEnglish = numberToText(1000000,'English');

  console.log('Converts 1000000 to ', textEnglish); //Converts 1000000 to One Million
```

```js
       
  //Large values

  var textIndian = numberToText(1000000000000);
  var textEnglish = numberToText(1000000000000,'English');

  console.log('Converts 1000000000000 to ', textIndian); //Converts 1000000000000 to One Lakh Crore
  console.log('Converts 1000000000000 to ', textEnglish); //Converts 1000000000000 to One Trillion
```

```js
       
  //For types other than Indian, English

  var text = numberToText(100,"Frensh");
  
  console.log('Retuns message: ', text); //Return message: Support for language: french is not available. Available languages are: indian,english
 ```

```js
       
  var defaultCurrency = numberToText(100.10,'',true);

  console.log('Converts 100.10 to ', defaultCurrency); //Converts 100.10 to One Hundred Rupee And Ten Paise Only
```

```js
       
  var defaultNonCurrency = numberToText(100.10);

  console.log('Converts 100.10 to ', defaultNonCurrency); //Converts 100.10 to One Hundred Point Ten 
```

## Executing tests

```shell
   npm test
```

## Contribution

Just need to take care to maintain the existing coding style. Add unit tests for any new or changed functionality. Lint and test your code.

```shell
  git clone https://github.com/garora/number2text.git
  cd number2text
```


# Starting with pre-release installation

```shell
  npm install git://github.com/garora/number2text.git
  npm install git://github.com/garora/number2text.git#pre-releaseversion
```

## License

MIT

## Release notes

Currently ```number2text``` is in initial release. We're trying to make it more stable. Currently, it supports Indian standard to read numbers as words. Check for more details: [Change log](https://github.com/garora/number2text/blob/master/Changelog.md)


## Release history

* 0.0.0 Initial release
* 0.1.0 Initial release - Fixed few issues related to build
* 0.1.1 Few fixes
* 0.1.11 Typos
* 1.0.0 New feature
* 1.0.1 Added large number support
* 1.1.0 Added currency support

Changelog.md
## 1.1.0 / March 11, 2015
  * Added currency support

## 1.0.1 / March 06, 2015
  * Added support for large number conversion
  

## 1.0.0 / March 06, 2015
  * Added functional and negative tests
  * Corrected logic to make performance
  * Added the ability so, it converts in English system of Million, Billion, Trillion etc.

## 0.1.11 / March 04, 2015
  * Beautify some stuffs
  
  
## 0.1.1 / March 04, 2015
  * Fixed few types with tests
  * Added more tests


## 0.1.0 / March 03, 2015
  * More functional tests
  * Added few more dev dependencies
  * Ability to show build status
  * Handling over/lower limit of entered numbers
  * Fixed [Bug](https://github.com/garora/number2text/issues/1)

## 0.0.0 / March 03, 2015
  * Initial release.


test/index.js
var should = require('chai').should(),
    number2text = require('../index');

 describe('ones', function() {
  
  it('Converts 1 to One', function() {
    number2text(1).should.equal('One');
    number2text(1,'english').should.equal('One');
    });

  it('Converts 2 to Two', function() {
    number2text(2).should.equal('Two');
    number2text(2,'english').should.equal('Two');
  });

  it('Converts 3 to Three', function() {
    number2text(3).should.equal('Three');
    number2text(3,'english').should.equal('Three');
  });

  it('Converts 4 to One', function() {
    number2text(4).should.equal('Four');
    number2text(4,'english').should.equal('Four');
  });

  it('Converts 5 to One', function() {
    number2text(5).should.equal('Five');
    number2text(5,'english').should.equal('Five');
  });

  it('Converts 6 to One', function() {
    number2text(6).should.equal('Six');
    number2text(6,'english').should.equal('Six');
  });

  it('Converts 7 to One', function() {
    number2text(7).should.equal('Seven');
    number2text(7,'english').should.equal('Seven');
  });

  it('Converts 8 to One', function() {
    number2text(8).should.equal('Eight');
    number2text(8,'english').should.equal('Eight');
  });

 it('Converts 9 to One', function() {
    number2text(9).should.equal('Nine');
    number2text(9,'english').should.equal('Nine');
  });

});

 describe('Tens', function() {
  
  it('Converts 10 to Ten', function() {
    number2text(10).should.equal('Ten');
    number2text(10,'english').should.equal('Ten');
  });

  it('Converts 20 to Twenty', function() {
    number2text(20).should.equal('Twenty');
    number2text(20,'english').should.equal('Twenty');
  });

  it('Converts 30 to Three', function() {
    number2text(30).should.equal('Thirty');
    number2text(30,'english').should.equal('Thirty');
  });

  it('Converts 40 to Fourty', function() {
    number2text(40).should.equal('Fourty');
    number2text(40,'english').should.equal('Fourty');
  });

  it('Converts 50 to Fifty', function() {
    number2text(50).should.equal('Fifty');
    number2text(50,'english').should.equal('Fifty');
  });

  it('Converts 60 to Sixty', function() {
    number2text(60).should.equal('Sixty');
    number2text(60,'english').should.equal('Sixty');
  });

  it('Converts 70 to Seventy', function() {
    number2text(70).should.equal('Seventy');
    number2text(70,'english').should.equal('Seventy');
  });

  it('Converts 80 to Eighty', function() {
    number2text(80).should.equal('Eighty');
    number2text(80,'english').should.equal('Eighty');
  });

  it('Converts 90 to Ninety', function() {
    number2text(90).should.equal('Ninety');
    number2text(90,'english').should.equal('Ninety');
  });
});

describe('Negative Tests', function() {
  
  it('Test for invalid number. N: Invalid number', function() {
    number2text('Hello').should.equal('Invalid number.');
    number2text('Hello','english').should.equal('Invalid number.');
  });

  it('Test for invalid number. S: in a string', function() {
  	var num = 's';
    number2text(num).should.equal('Invalid number.');
    number2text(num,'english').should.equal('Invalid number.');
  });

 it('Out of range - Lower', function() {
    number2text(-1).should.equal('Please enter +ve number only.');
    number2text(-1,'english').should.equal('Please enter +ve number only.');
  });

});

describe('Functional Tests', function() {
  
  it('Converts 9001 to Nine Thousand One', function() {
    number2text(9001).should.equal('Nine Thousand One');
    number2text(9001,'english').should.equal('Nine Thousand One');
  });
	
	it('Should converts 999999999 to its supplied type', function() {
    number2text(999999999).should.equal('Ninety-Nine Crore Ninety-Nine Lakh Ninety-Nine Thousand Nine Hundred Ninety-Nine');
    number2text(999999999,'english').should.equal('Nine Hundred Ninety-Nine Million Nine Hundred Ninety-Nine Thousand Nine Hundred Ninety-Nine');
  });


	it('Converts 10000000 to One Crore', function() {
    number2text(10000000).should.equal('One Crore');
    number2text(10000000,'english').should.equal('Ten Million');
  });

	it('Converts 100000000 to Ten Crore', function() {
    number2text(100000000).should.equal('Ten Crore');
    number2text(100000000,'english').should.equal('One Hundred Million');
  });

	it('Converts 1000000 to One Lakh', function() {
    number2text(100000).should.equal('One Lakh');
    number2text(100000,'english').should.equal('One Hundred Thousand');
  });

	it('Converts 1000000 to Ten Lakh', function() {
    number2text(1000000).should.equal('Ten Lakh');
    number2text(1000000,'english').should.equal('One Million');
  });

	it('Converts 10000 to Ten Thousand', function() {
    number2text(10000).should.equal('Ten Thousand');
    number2text(10000,'english').should.equal('Ten Thousand');
  });
	it('Converts 1000 to One Thousand', function() {
    number2text(1000).should.equal('One Thousand');
    number2text(1000,'english').should.equal('One Thousand');
  });

	it('Converts 100 to One Hundred', function() {
    number2text(100).should.equal('One Hundred');
    number2text(100,'english').should.equal('One Hundred');
  });

	it('Converts 0 to Zero', function() {
    number2text(0).should.equal('Zero');
  	number2text(0,'english').should.equal('Zero');
  });
});


describe('Conversion Language Tests', function() {
	
	it('Default conversion type is Indian (100000 to One Lakh)', function() {
	    number2text(100000).should.equal('One Lakh');
	    number2text(100000,'').should.equal('One Lakh');
	    number2text(100000).should.equal(number2text(100000,'Indian'));
	  });

	it('English conversion for (1000000 to One Lakh)', function() {
	    number2text(10000000,'english').should.equal('Ten Million');
	  });

	it('French conversion is not available', function() {
	    number2text(1,'french').should.equal('Support for type: french is not available. Available types are: indian,english');
	  });
});

describe('Extra large numbers', function() {
	
	it('Should convert extra large numbers', function() {
	    number2text(1000000000).should.equal('One Hundred Crore');
	    number2text(1000000000,'English').should.equal('One Billion');

	    number2text(1000000000000).should.equal('One Lakh Crore');
	    number2text(1000000000000,'English').should.equal('One Trillion');

	    number2text(1000000000000000).should.equal('Ten Crore Crore'); 
	    number2text(1000000000000000,'English').should.equal('One Quadrillion');

	    number2text(1000000000000000000).should.equal('Ten Thousand Crore Crore'); //High number discrepency in Indian type
	    number2text(1000000000000000000,'English').should.equal('One Quintillion');

	    number2text(10000000000000000000).should.equal('One Lakh Crore Crore'); //High number discrepency in Indian type
	    number2text(10000000000000000000,'English').should.equal('Ten Quintillion');
	  });
});

describe('Decimal Tests', function() {
  
 it('1010.10 should convert to One Thousand Ten Point Ten with default type', function() {
    number2text('1010.10').should.equal('One Thousand Ten Point Ten');
  });

 it('1010.10 should convert to One Thousand Ten Point Ten with type english', function() {
    number2text('1010.10','english').should.equal('One Thousand Ten Point Ten');
  });

 it('Should validate correct type', function() {
    number2text('1010.10',true).should.equal('Please enter valid type.');
  });

 it('Should handle decimal', function() {
    number2text('1010.10','indian').should.equal(number2text('1010.10'));
  });

 it('Should handle decimal using default type', function() {
    number2text('1010.10','',true).should.equal(number2text('1010.10','indian',true));
  });

 it('Should validate types using decimal', function() {
    number2text('1010.10','indian',true).should.not.equal(number2text('1010.10','english',true));
  });

});

describe('Currency Tests', function() {
  
 it('1010.10 should convert to One Thousand And  Ten Rupee And Ten Paise only with default type', function() {
    number2text('1010.10','',true).should.equal('One Thousand And  Ten Rupee And Ten Paise only');
  });

 it('Should handle currency support with default type', function() {
    number2text('1010.10','',false).should.not.equal(number2text('1010.10','',true));
  });

 it('1010.10 should convert to One Thousand And  Ten Dollar And Ten Cent only with type english', function() {
    number2text('1010.10','english',true).should.equal('One Thousand And  Ten Dollar And Ten Cent only');
  });

 it('Should handle currency support with default type', function() {
    number2text('1010.10','english',false).should.not.equal(number2text('1010.10','english',true));
  });

});

lib/numberToText.js
/**
* Note: This file require refactoring
*/

function numberToText(num, language,isCurrency)
{
  var lang = new Array("indian", "english"); //make in lower case
  var stringValue = "This is a String";
  

  if(isNaN(num))
     return "Invalid number.";
  
  if(num < 0)
    return "Please enter +ve number only.";
    
  if(isBlank(language) || isNull(language))
    language = "indian";

   if (typeof language != typeof stringValue)
    return "Please enter valid type.";
    
      
  if(lang.indexOf(language.toLowerCase()) == -1)
        return "Support for type: " + language + " is not available. Available types are: " + lang.toString();
       
  return toText(num,language,isCurrency); //currently supports for Indian number conversions    
}


function isNull(varIsNull){  return (typeof varIsNull === 'undefined'); }

function isBlank(str) {    return (!str || /^\s*$/.test(str));}

function toText(num,type,isCurrency){  
    
    var res;
    var fract_part = Math.round(frac_one(num)*100);
    var fract_num  = "";

    switch (type.toLowerCase()) {
        
        case "indian":           
            if(fract_part > 0) 
            {            
                if(isCurrency)
                    fract_num =  "And "+ toIndianText(fract_part,isCurrency) + " Paise" ;
                else
                    fract_num =  " Point "+ toIndianText(fract_part,isCurrency);
            }

            if(isCurrency)
                res = toIndianText(num,isCurrency) + " Rupee " + fract_num+" only";
            else
                res = toIndianText(num,isCurrency) + fract_num;
            break;
        case "english":
            if(fract_part > 0) 
            {            
                if(isCurrency)
                    fract_num =  "And "+ toEnglishText(fract_part,isCurrency) + " Cent" ;
                else
                    fract_num =  " Point "+ toEnglishText(fract_part,isCurrency);
            }

            if(isCurrency)
                res = toEnglishText(num,isCurrency) + " Dollar " + fract_num+" only";
            else
                res = toEnglishText(num,isCurrency) + fract_num;

            break;
    }
    return res;
}

function frac_one(num) {return num % 1;}

function toEnglishText(num,isCurrency)
{
    //For English conversion

    var quintillion = Math.floor(num / 1000000000000000000); /* quintillion */
    num -= quintillion * 1000000000000000000;
    var quar = Math.floor(num / 1000000000000000); /* quadrillion */
    num -= quar * 1000000000000000;
    var trin = Math.floor(num / 1000000000000); /* trillion */
    num -= trin * 1000000000000;
    var Gn = Math.floor(num / 1000000000); /* billion */
    num -= Gn * 1000000000;
    var million = Math.floor(num / 1000000); /* million */
    num -= million * 1000000;
    var Hn = Math.floor(num / 1000); /* thousand */
    num -= Hn * 1000;
    var Dn = Math.floor(num / 100); /* Tens (deca) */
    num = num % 100; /* Ones */
    var tn = Math.floor(num / 10);
    var one = Math.floor(num % 10);
    var res = "";

    if (quintillion > 0) {
        res += (toEnglishText(quintillion) + " Quintillion");
    }
    if (quar > 0) {
        res += (toEnglishText(quar) + " Quadrillion");
    }
    if (trin > 0) {
        res += (toEnglishText(trin) + " Trillion");
    }
    if (Gn > 0) {
        res += (toEnglishText(Gn) + " Billion");
    }
    if (million > 0) {
        res += (((res == "") ? "" : " ") + toEnglishText(million) + " Million");
    }
    if (Hn > 0) {
        res += (((res == "") ? "" : " ") + toEnglishText(Hn) + " Thousand");
    }

    if (Dn) {
        res += (((res == "") ? "" : " ") + toEnglishText(Dn) + " Hundred");
    }


    var ones = Array("", "One", "Two", "Three", "Four", "Five", "Six", "Seven", "Eight", "Nine", "Ten", "Eleven", "Twelve", "Thirteen", "Fourteen", "Fifteen", "Sixteen", "Seventeen", "Eightteen", "Nineteen");
    var tens = Array("", "", "Twenty", "Thirty", "Fourty", "Fifty", "Sixty", "Seventy", "Eighty", "Ninety");

    if (tn > 0 || one > 0) {
        if(isCurrency)
        {
            if (!(res == "")) {
            res += " And "; //for currency only
            }
        }

        if (!(res == "")) {
            res += " ";
        }


        if (tn < 2) {
            res += ones[tn * 10 + one];
        } else {

            res += tens[tn];
            if (one > 0) {
                res += ("-" + ones[one]);
            }
        }
    }

    if (res == "") {
        res = "Zero";
    }
    return res;
}

//This function is used for Indian conversion only
function toIndianText(num,isCurrency){
    var Gn = Math.floor(num / 10000000);  /* Crore */ 
    num -= Gn * 10000000; 
    var kn = Math.floor(num / 100000);     /* lakhs */ 
    num -= kn * 100000; 
    var Hn = Math.floor(num / 1000);      /* thousand */ 
    num -= Hn * 1000; 
    var Dn = Math.floor(num / 100);       /* Tens (deca) */ 
    num = num % 100;               /* Ones */ 
    var tn= Math.floor(num / 10); 
    var one=Math.floor(num % 10); 

    var res = ""; 

    if (Gn>0) 
    { 
        res += (toIndianText(Gn) + " Crore"); 
    } 
    if (kn>0) 
    { 
            res += (((res=="") ? "" : " ") + 
            toIndianText(kn) + " Lakh"); 
    } 
    if (Hn>0) 
    { 
        res += (((res=="") ? "" : " ") +
            toIndianText(Hn) + " Thousand"); 
    } 

    if (Dn) 
    { 
        res += (((res=="") ? "" : " ") + 
            toIndianText(Dn) + " Hundred"); 
    } 


    var ones = Array("", "One", "Two", "Three", "Four", "Five", "Six","Seven", "Eight", "Nine", "Ten", "Eleven", "Twelve", "Thirteen","Fourteen", "Fifteen", "Sixteen", "Seventeen", "Eighteen","Nineteen"); 
    var tens = Array("", "", "Twenty", "Thirty", "Fourty", "Fifty", "Sixty","Seventy", "Eighty", "Ninety");

    if (tn>0 || one>0) 
    { 
          if(isCurrency)
          {
              if (!(res=="")) 
                res += " And ";
          }

        if (!(res=="")) 
                res += " ";

        if (tn < 2) 
        { 
            res += ones[tn * 10 + one]; 
        } 
        else 
        { 

            res += tens[tn];
            if (one>0) 
            { 
                res += ("-" + ones[one]); 
            } 
        } 
    }

    if (res=="")
    { 
        res = "Zero"; 
    } 
    return res;

}


//export
module.exports = numberToText;
